cdap-master/src/main/java/io/cdap/cdap/master/environment/k8s/AuthenticationServiceMain.java

```diff
@@ -0,0 +1,152 @@
+import org.apache.twill.zookeeper.ZKClientService;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.util.Arrays;
+import java.util.List;
+import java.util.concurrent.TimeUnit;
+
+/**
+ * The main class responsible for Authentication  .
+ */
+public class AuthenticationServiceMain extends AbstractServiceMain<EnvironmentOptions> {
+
+  private static final Logger LOG = LoggerFactory.getLogger(AuthenticationServiceMain.class);
+
+  public static void main(String[] args) throws Exception {
+    main(AuthenticationServiceMain.class, args);
+  }
+
+  @Override
+  protected List<Module> getServiceModules(
+      MasterEnvironment masterEnv, EnvironmentOptions options, CConfiguration cConf) {
+    return Arrays.asList(
+        getDataFabricModule(),
+        new ZKClientModule(),
+        new SecurityModules().getDistributedModules(),
+        new MessagingClientModule());
+  }
+
+  @Override
+  protected void addServices(
+      Injector injector, List<? super Service> services, List<? super AutoCloseable> closeableResources,
+      MasterEnvironment masterEnv, MasterEnvironmentContext masterEnvContext, EnvironmentOptions options) {
+
+    MessagingService messagingService = injector.getInstance(MessagingService.class);
+    if (messagingService instanceof Service) {
+      services.add((Service) messagingService);
+    }
+
+    CConfiguration configuration = injector.getInstance(CConfiguration.class);
+
+    if (configuration.getBoolean(Constants.Security.ENABLED)) {
+
+      services.add(new AbstractService() {
+
+        @Override
+        protected void doStart() {
+
+          ZKClientService zkClientService = injector.getInstance(ZKClientService.class);
+          ExternalAuthenticationServer authServer = injector.getInstance(ExternalAuthenticationServer.class);
+
+          try {
+            LOG.info("Starting AuthenticationServer.");
+
+            // Enable Kerberos login
+            SecurityUtil.enableKerberosLogin(configuration);
+
+            io.cdap.cdap.common.service.Services.startAndWait(
+                zkClientService,
+                configuration.getLong(
+                    Constants.Zookeeper.CLIENT_STARTUP_TIMEOUT_MILLIS),
+                TimeUnit.MILLISECONDS,
+                String.format(
+                    "Connection timed out while trying to start " +
+                        "ZooKeeper client. Please verify that the " +
+                        "ZooKeeper quorum settings are correct in " +
+                        "cdap-site.xml. Currently configured as: %s",
+                    configuration.get(Constants.Zookeeper.QUORUM)));
+            authServer.startAndWait();
+
+            notifyStarted();
+          } catch (Exception e) {
+            Throwable rootCause = Throwables.getRootCause(e);
+            if (rootCause instanceof ServiceBindException) {
+              LOG.error("Failed to start Authentication Server: {}", rootCause.getMessage());
```

Comment:
So if it fails to connect to ZK, there is just logging but not failing the pod? Is that an intended behavior?

Reply:
Added System.exit(1).